cranelift/codegen/src/isa/x64/lower.isle

```diff
@@ -1970,18 +1963,11 @@
 ;; Analogous to `clz` cases above, but using mirror instructions
 ;; (tzcnt vs lzcnt, bsf vs bsr).
 
-(rule 2 (lower
-         (has_type (and
-                    (ty_32_or_64 ty)
-                    (use_bmi1 $true))
-                   (ctz src)))
+(rule 3 (lower (has_type (ty_32_or_64 ty) (ctz src)))
+      (if-let $true (use_bmi1))
       (x64_tzcnt ty src))
 
-(rule 2 (lower
-          (has_type (and
-                     (ty_32_or_64 ty)
-                     (use_bmi1 $false))
-                 (ctz src)))
+(rule 2 (lower (has_type (ty_32_or_64 ty) (ctz src)))
```

Comment:
I have a preference for using pattern-matches rather than priorities whenever possible.

In this case I could be convinced the other way: do_ctz is more general and still works even if use_bmi1 is enabled, so semantically the priorities make sense; and it probably doesn't make much difference in terms of LLVM's output.

But in general, priorities are kinda weird action-at-a-distance, and they can impede ISLE optimizations.

Reply:
ISLE used to treat more "specific" rules as implicitly higher priority, but we've been moving away from this kind of implicit priority scheme because it was hard for rule authors to reason about. We have all the code needed to implement it in ISLE if we wanted to, though; it's just an application of the subset-overlap check.

If you use `if-let` in both rules, matching against `$true` and `$false` explicitly, you'll get good codegen without priorities, and the overlap check will understand that the two rules are mutually exclusive. But as I said that's a little unfortunate here where the `$false` case would still be correct if the `$true` case is removed. So maybe this is just fine as-is.

Reply:
Yeah I realize the implicit priority scheme had its own issues but my hope was that "zero `if-let`" vs "one or more `if-let`" is pretty clear in terms of priority to not need an explicit disambiguation. So not the full "which `if-let` is more specific", just a zero-vs-nonzero check. My hope is that the idiom of "write a rule with `if-let`" ideally wouldn't require explicit priorities and could be handled in the common case because it's nice to not have to update all other lowerings when a more explicit one is added.